libcudacxx/include/cuda/__floating_point/operators.h

```diff
@@ -0,0 +1,1250 @@
+struct __fp_ops
+{
+  /********************************************************************************************************************/
+  // Negation
+  /********************************************************************************************************************/
+  template <class _Tp>
+  _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI constexpr static _Tp __neg_impl_constexpr(const _Tp& __src) noexcept
+  {
+    auto __ret{__src};
+    __ret.__set_sign(!__ret.__get_sign());
+    return __ret;
+  }
+
+  template <class _Tp>
+  _CCCL_NODISCARD _CCCL_HIDE_FROM_ABI _CCCL_DEVICE static _Tp __neg_impl_device(const _Tp& __src) noexcept
+  {
+    [[maybe_unused]] _Tp __ret;
+
+#  if __cccl_ptx_isa >= 600
+    if constexpr (_CUDA_VSTD::_CCCL_TRAIT(is_same, _Tp, fp16))
+    {
+      NV_IF_TARGET(NV_PROVIDES_SM_53,
+                   (asm("neg.f16 %0, %1;" : "=h"(__ret.__storage_) : "h"(__src.__storage_)); return __ret;))
+    }
+#  endif // __cccl_ptx_isa >= 600
+#  if __cccl_ptx_isa >= 700
+    if constexpr (_CUDA_VSTD::_CCCL_TRAIT(is_same, _Tp, bf16))
+    {
+      NV_IF_TARGET(NV_PROVIDES_SM_80,
+                   (asm("neg.bf16 %0, %1;" : "=h"(__ret.__storage_) : "h"(__src.__storage_)); return __ret;))
+    }
+#  endif // ^^^ __cccl_ptx_isa < 700 ^^^
+
+    return __neg_impl_constexpr(__src);
+  }
+
+  template <class _Tp>
+  _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI static constexpr _Tp __neg_impl(const _Tp& __src) noexcept
+  {
+    if (!_CUDA_VSTD::__cccl_default_is_constant_evaluated())
+    {
+      NV_IF_TARGET(NV_IS_DEVICE, (return __neg_impl_device(__src);))
+    }
+
+    return __neg_impl_constexpr(__src);
+  }
+
+  template <class _Tp>
+  _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI static constexpr auto __neg(const _Tp& __src) noexcept
+  {
+    NV_IF_ELSE_TARGET(
+      NV_IS_HOST,
+      (
+        if constexpr (_Tp::__has_host_native_type) { return __fp{__fp_from_native, -__src.__host_native()}; } else {
+          return __neg_impl(__src);
+        }),
+      (
+        if constexpr (_Tp::__has_device_native_type) { return __fp{__fp_from_native, -__src.__device_native()}; } else {
+          return __neg_impl(__src);
+        }))
+  }
+
+  /********************************************************************************************************************/
+  // Addition
+  /********************************************************************************************************************/
+  template <class _Lhs, class _Rhs>
+  _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI constexpr static __fp_common_type_t<_Lhs, _Rhs>
+  __add_impl_constexpr(const _Lhs& __lhs, const _Rhs& __rhs) noexcept
+  {
+    if constexpr ((_CUDA_VSTD::is_same_v<_Lhs, fp16> && _CUDA_VSTD::is_same_v<_Rhs, fp16>)
+                  || (_CUDA_VSTD::is_same_v<_Lhs, bf16> && _CUDA_VSTD::is_same_v<_Rhs, bf16>) )
+    {
+      return _Lhs{static_cast<float>(__lhs) + static_cast<float>(__rhs)};
```

Comment:
This needs to take care that the rounding due to the round trip is right, compared to native arithmetic. Also, this code should really be more generic in how it looks for the native type to use, but I think you know that :)